cli/pcluster/config/param_types.py

```diff
@@ -0,0 +1,1202 @@
+
+        :param config_parser: the configparser object from which get the parameter
+        :raise NoOptionError, NoSectionError if unable to get the param from config_parser
+        """
+        section_name = _get_file_section_name(self.section_key, self.section_label)
+        self.value = config_parser.get(section_name, self.key)
+        self._check_allowed_values()
+
+    def _init_from_cfn(self, cfn_params=None, cfn_value=None):
+        """
+        Initialize parameter value by parsing CFN input parameters or from a given value coming from CFN.
+
+        :param cfn_params: the list of all the CFN parameters, it is used if "cfn" attribute is specified in the map
+        :param cfn_value: a value coming from a comma separated CFN param
+        """
+        cfn_converter = self.map.get("cfn", None)
+        if cfn_converter and cfn_params:
+            cfn_value = get_cfn_param(cfn_params, cfn_converter) if cfn_converter else "NONE"
+            self.value = self.get_value_from_string(cfn_value)
+        elif cfn_value:
+            self.value = self.get_value_from_string(cfn_value)
+        else:
+            self._init_from_map()
+
+    def _init_from_map(self):
+        """Initialize parameter value by using default specified in the mapping file."""
+        self.value = self.get_default_value()
+        if self.value:
+            LOGGER.debug("Setting default value '%s' for key '%s'", self.value, self.key)
+
+    def _check_allowed_values(self):
+        """Verify if the parameter value is one of the allowed values specified in the mapping file."""
+        allowed_values = self.map.get("allowed_values", None)
+        if allowed_values:
+            if isinstance(allowed_values, list):
+                if self.value not in allowed_values:
+                    error(
+                        "The configuration parameter '{0}' has an invalid value '{1}'\n"
+                        "Allowed values are: {2}".format(self.key, self.value, allowed_values)
+                    )
+            else:
+                # convert to regex
+                if not re.compile(allowed_values).match(str(self.value)):
+                    error(
+                        "The configuration parameter '{0}' has an invalid value '{1}'\n"
+                        "Allowed values are: {2}".format(self.key, self.value, allowed_values)
+                    )
+
+    def validate(self, fail_on_error=True):
+        """Call validation function for the parameter, if there."""
+        validation_func = self.map.get("validator", None)
+
+        if not validation_func:
+            LOGGER.debug("Configuration parameter '%s' has no validator", self.key)
+        elif self.value is None:
+            LOGGER.debug("Configuration parameter '%s' has not a value", self.key)
+        else:
+            errors, warnings = validation_func(self.key, self.value, self.pcluster_config)
+            if errors:
+                error(
```

Comment:
It is a good idea and I like it a lot, but I have to change a lot of tests after this change because the code won't raise SystemExit when the value is wrong

I'd leave it for a next patch in this way we can go ahead and merge this big patch. 
In any case we are not changing the old behaviour on which the code was fast failing in the same way.